Image - support blurry loading

nitayneeman opened the pull request:
This PR adds a blurry progressive loading for `<Image/>` inspired by [Restaurants](https://github.com/wix-private/restaurants/blob/master/restaurants-olo-client/src/components/MainPage/components/PreviewedImage/PreviewedImage.tsx#L100).

Notice it's not the exact implementation, but the concept is pretty similar.

nitayneeman commented on src/components/Image/Image.tsx:
```diff
@@ -17,37 +17,106 @@ export interface ImageProps extends TPAComponentProps {
+  loadingBehavior?: 'none' | 'blur';
 }
 
+interface Dimensions {
+  width: ImageProps['width'];
+  height: ImageProps['height'];
+}
```
I tried to use `Pick` but the problem is that `width` & `height` are optional at the moment in `ImageProps`. Either way, the dimensions logic is probably going to change...

nitayneeman commented on src/components/Image/Image.visual.tsx:
```diff
@@ -48,6 +48,15 @@ visualize('Image', () => {
+      snap('with blurry loading', (done) => (
+        <Image
+          src={src}
+          width={200}
+          height={200}
```
Currently we have to pass the dimensions to render a media image, so we cannot take the screenshot without them (even though they are not related to the test)

nitayneeman commented:
I'm reviewing the animation with @michalvardi

WixBuildServer commented:
### 🤖 Hi, I'm #Janet!

  A deploy preview of your PR was successfully released 🖼.

  You can find it under one of the following URLs:
  * https://bo.wix.com/pages/wix-ui-tpa.pr_567/

shlomitc commented on src/components/Image/Image.st.css:
```diff
@@ -1,2 +1,32 @@
+  display: grid;
+}
+
+.root img {
```
why don't you use the `MediaImage` style API?

nitayneeman replied:
I'm not sure there is a style API...

These are the props:
https://github.com/wix/wix-ui/blob/master/packages/wix-ui-core/src/components/media-image/media-image.tsx#L96-L102

I thought about exposing `className` prop - WDYT?

jonathanadler replied:
Yeah I think a `className` prop is definitely needed no matter the approach.
That way you won't need to target any internal elements.

I would even add it to wix-ui-core before merging this PR.

nitayneeman replied:
I'll add a `className` for the wrapper div in `Image` & `MediaImage`(since I use both, depends on `src`), but I'll still need to target the native `img` inside.

I can do that in two ways:
1. Expose a prop like `nativeElementClass` on the native `img`
2. Make the native `img` extensible using Stylable

Any preference?

shlomitc commented on src/components/Image/docs/examples.ts:
```diff
@@ -17,3 +17,13 @@ export const relativeUriExample = `
 />
 `;
+
+export const blurLoadingExample = `
```
consider adding a button to retrigger the rendering since it's very quick.

nitayneeman replied:
Basically you can just change any value in the props to see it again, but I'll add a button to make it easier 👌

jonathanadler commented on src/components/Image/Image.st.css:
```diff
@@ -1,2 +1,32 @@
+  grid-area: 1 / 1;
+}
+
+.root:preload img:last-child {
+  filter: blur(value(BlurRadius));
+}
+
+.root:loaded img:last-child {
+  animation-name: blurry-loading;
+  animation-duration: 0.3s;
+  animation-timing-function: easeout;
 }
```
After adding the `className` prop to the MediaImage component, I would give different class names to the two images, just so not to rely on the dom ordering, and for better readability.

jonathanadler commented on src/components/Image/Image.tsx:
```diff
@@ -17,37 +17,106 @@ export interface ImageProps extends TPAComponentProps {
 
+  state = { isLoaded: false };
+
+  _onLoad(event) {
```
```suggestion
  _onLoad = (event) => {
```

jonathanadler commented on src/components/Image/Image.tsx:
```diff
@@ -17,37 +17,106 @@ export interface ImageProps extends TPAComponentProps {
             {...imageProps}
             nativeProps={{ ...dimensions }}
             src={src}
+            onLoad={(event) => this._onLoad(event)}
```
```suggestion
            onLoad={this._onLoad}
```

jonathanadler reviewed (commented):
Nice!

Left some small comments.

nitayneeman commented:
@jonathanadler I added `className` props in ui-core:
https://github.com/wix/wix-ui/pull/2000

And also changed the component implementation to use them, I agree it's much readable than before.

jonathanadler reviewed (approved):
The only thing I'm not sure about, is how to reflect the fact that using the blur load with an absolute src is merely visual, and doesn't have an affect on performance.

nitayneeman commented:
> The only thing I'm not sure about, is how to reflect the fact that using the blur load with an absolute src is merely visual, and doesn't have an affect on performance.

I'll mention it on the story page.
